src/proc_underwater_com_node.cc

```diff
@@ -46,66 +46,181 @@ namespace proc_underwater_com
         syncPublisher_ =  nh_->advertise<std_msgs::Bool>("/proc_underwater_com/sync_requested", 100);
         DepthPublisher_ =  nh_->advertise<std_msgs::Float32>("/proc_underwater_com/other_sub_depth", 100);
 
-        // Service
+        // Services
         depthSrv_ = nh_->advertiseService("/proc_underwater_com/depth_request", &ProcUnderwaterComNode::DepthRequest, this);
         underwaterComClient_ = nh_->serviceClient<sonia_common::ModemSendCmd>("/provider_underwater_com/request");
 
+        // Threads
+        interpretMessage_ = std::thread(std::bind(&ProcUnderwaterComNode::UnderwaterComInterpreter, this));
+        sendMessageToSensor_ = std::thread(std::bind(&ProcUnderwaterComNode::SendMessageToSensor, this));
+
         InitMissionState(configuration_.getNumberMission());
         AUVID = configuration_.getNumberid();
+
+        ackknowledge_mission_completed_ = 0;
+        ackknowledge_sync_completed_ = 0;
     }
 
     // Node Destructor
-    ProcUnderwaterComNode::~ProcUnderwaterComNode(){}
+    ProcUnderwaterComNode::~ProcUnderwaterComNode()
+    {
+        stop_thread = true;
+    }
 
     // Spin
     void ProcUnderwaterComNode::Spin()
     {
-        ros::Rate r(5); // 5 Hz
+        ros::Rate r(20); // 20 Hz
 
         while(ros::ok())
         {
             ros::spinOnce();
             r.sleep();
         }
     }
-
     void ProcUnderwaterComNode::UnderwaterComInterpreterCallback(const std_msgs::UInt64 &msg)
     {
-        Modem_M64_t packet = ConstructPacket(msg.data);
+        ROS_DEBUG_STREAM("Received message");
+        queue.push_back(msg.data);
+    }
+
+    void ProcUnderwaterComNode::UnderwaterComInterpreter()
+    {
+        Modem_M64_t packet;
         float_t auvDepth = 0;
         uint32_t temp = 0;
-        uint8_t data[] = {packet.data[0], packet.data[1], packet.data[2], packet.data[3]};
+        uint8_t read_write = 0;
+        uint8_t data[] = {0,0,0,0,0,0};
+
+        while(!stop_thread)
+        {
+            if(!queue.empty())
```

Comment:
if I'm not wrong, this is a blocking queue so you can just wait on the pop

Reply:
We will have to try it out.

Reply:
@alecst-marseille Start with this one